### DISCUSSION TITLE
build: Add Support for Nix Packaging on Darwin (macOS)

### DISCUSSION THREAD

**Discussion by r17x** [ghostty-org/ghostty#2824] [Q&A] [answered]
## Description
The current Nix packaging setup is designed exclusively for Linux, and there is no clear support for Darwin (macOS). As a macOS user, I’d like to contribute to the project and streamline the development process on macOS. Supporting Nix packaging on Darwin will ensure a consistent development environment across platforms and make it easier for macOS users to contribute effectively.

## Questions
* Is adding Darwin support within the scope of this project?
* Are there existing efforts or blockers for macOS compatibility that I should be aware of?
* Would the maintainers accept contributions to add Darwin support?

### Current Behavior

```console
❯ nix build
```

* The Nix expressions and packages are tailored for Linux.
* Attempts to build the environment on macOS result in compatibility issues or unsupported configurations.

Environment
```console
❯ nix-info -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.1.0, macOS 15.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source`
```
### Notes

I’m happy to assist in testing and contributing to this effort if support for Darwin aligns with the project’s goals. Let me know how I can help!

**Comment by mitchellh** [comment, marked as answer]
Thanks for asking the questions. This gets raised periodically and I think having a single online resource I can point to with the answers will be helpful. I'll use this issue/discussion for that. 😄 

> Is adding Darwin support within the scope of this project?

_Kind of_. 

The general stance I have is that _packaging_ is out of scope for this project. Packaging will be up to the community. The one exception (which is valid at the time of writing) is during the private beta it is infeasible for many platforms to create packaging to reference closed source or private repos. So, definitely once Ghostty is public, packaging isn't the responsibility of the project.

There are a couple exceptions there. One is macOS in general: since macOS users expect binary, signed, and notarized releases with auto-updating, I plan on supporting the official macOS binary release channels. That's sort of idiomatic macOS. Windows (in the future) will probably be the same. So the above primarily applies to Linux due to the distro diversity and cultural differences.

Second is Nix. Nix is the _only officially supported dev/build environment_ for the project. So I plan on keeping at least our `devShell` working. The `package` part of our flake is how we test general Linux packaging. 

So, let's get to the "kind of" answer: I'd be willing to accept it under the conditions we can keep CI passing for it. But it wouldn't be an official release channel.

> Are there existing efforts or blockers for macOS compatibility that I should be aware of?

Yes, many, but they're unfortunately locked away in Discord history. Every few months some new beta person has been invited and decided to try to take this on, most of them much better and more experienced at Nix than me. 😄 

As far as I know, Nix support for building macOS app bundles is lacking and everyone hits some kind of road block. I don't know the details, I'm sorry.

> Would the maintainers accept contributions to add Darwin support?

Yes, under the conditions stated above. Namely that it can be tested in CI, but also I don't want anything hacky (forcing the tooling), and that it won't be an officially supported release channel. I would also recommend -- especially when public -- the community maintaining an external flake to run this.

I will add that I'm also open to adding changes to `build.zig` that are reasonable to support this. One that has been brought up before has been the ability to build non-universal binaries. Right now, our build on macOS forces a lipo x86/arm64 libghostty build. I'm okay with adding a non-default build option to create only a single-architecture target.

I hope this helps, and good luck! I'm happy to answer any followups.

**Comment by garbas** [comment]
Somebody already started working on this?

**Reply by r17x** [reply]
I have been trying in my local and have same issue (I'm skill issue for building MacOS Apps over nix)

**Reply by Eveeifyeve** [reply]
I have been working on it as [this nixpkgs pr](https://github.com/NixOS/nixpkgs/pull/368404) just came out today, it doesn't have darwin support yet.. but I am planning to add it but the zig is not giving me much verbose what is wrong with the apple_sdk package in this repo.

**Reply by Eveeifyeve** [reply]
I just think it's not finding the apple sdk in this repo I don't know if this repo issue or nixpkgs but It could be looking like and issue with the `pkgs/apple_sdk` not finding the apple sdk that nix provides.

**Reply by rhodes-b** [reply]
https://discord.com/channels/1005603569187160125/1320424792461152296
and 
https://discord.com/channels/1005603569187160125/1321955142941020260
if you are in the discord

**Comment by jaycem-dev** [comment]
Apple just made their Xcode build engine [Open Source](https://www.swift.org/blog/the-next-chapter-in-swift-build-technologies/). I'm not that familiar with Swift, but it seems like it could help with this.

**Comment by nstrelow** [comment]
Is this acutally achievable at some point? Ghostty is my only brew dep currently, would like to make it fully nix ^^

**Reply by pluiedev** [reply]
Nixpkgs doesn't even have Swift 6. It'll be a *long* time before this will become feasible.

**Reply by nstrelow** [reply]
Gotcha, thanks for the reply.

I am praying for the day of complete purity to come 🙌
